src/game/logic/object/experiencetracker.cpp

```diff
@@ -0,0 +1,181 @@
+/**
+ * @file
+ *
+ * @author Duncans_pumpkin
+ *
+ * @brief ExperienceTracker
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "experiencetracker.h"
+#include "gamelogic.h"
+#include "object.h"
+#include "xfer.h"
+
+// zh: 0x0061B790 wb: 0x008A35F9
+int32_t ExperienceTracker::Get_Experience_Value(const Object *killer) const
+{
+    if (killer->Get_Relationship(m_parent) == Relationship::ALLIES) {
+        return 0;
+    }
+
+    return m_parent->Get_Template()->Get_Experience_Value(m_currentLevel);
+}
+
+// zh: 0x0061B7D0 wb: 0x008A3637
+bool ExperienceTracker::Is_Trainable() const
+{
+    return m_parent->Get_Template()->Is_Trainable();
+}
+
+// zh: 0x0061B7E0 wb: 0x008A3654
+bool ExperienceTracker::Is_Accepting_Experience_Points() const
+{
+    return Is_Trainable() || m_experienceSink != ObjectID::OBJECT_UNK;
+}
+
+// zh: 0x0061B820 wb: 0x008A36A4
+void ExperienceTracker::Set_Min_Veterency_Level(VeterancyLevel newLevel)
+{
+    if (m_currentLevel >= newLevel) {
+        return;
+    }
+
+    captainslog_dbgassert(m_parent != nullptr, "Parent must not be nullptr!");
+
+    const auto old_level = m_currentLevel;
+    m_currentLevel = newLevel;
+
+    m_currentExperience = m_parent->Get_Template()->Get_Experience_Required(m_currentLevel);
+
+    m_parent->On_Veterancy_Level_Changed(old_level, m_currentLevel, true);
+}
+
+// zh: 0x0061B860 wb: 0x008A370D
+void ExperienceTracker::Set_Veterency_Level(VeterancyLevel newLevel, bool unk)
+{
+    if (m_currentLevel == newLevel) {
+        return;
+    }
+
+    captainslog_dbgassert(m_parent != nullptr, "Parent must not be nullptr!");
+
+    const auto old_level = m_currentLevel;
+    m_currentLevel = newLevel;
+
+    m_currentExperience = m_parent->Get_Template()->Get_Experience_Required(m_currentLevel);
+
+    m_parent->On_Veterancy_Level_Changed(old_level, m_currentLevel, true);
+}
+
+// zh: 0x0061B8A0 wb: 0x008A377A
+bool ExperienceTracker::Gain_Exp_For_Level(int32_t levelsToGain, bool applyMultiplier)
+{
+    const auto new_level = std::min<VeterancyLevel>(m_currentLevel + levelsToGain, VeterancyLevel::VETERANCY_COUNT - 1);
+
+    if (new_level <= m_currentLevel) {
+        return false;
+    }
+
+    const auto exp_for_level_gain = m_parent->Get_Template()->Get_Experience_Required(new_level) - m_currentExperience;
+
+    Add_Experience_Points(exp_for_level_gain, applyMultiplier);
+    return true;
+}
+
+// zh: 0x0061B8F0 wb: 0x008A37E4
+bool ExperienceTracker::Can_Gain_Exp_For_Level(int32_t levelsToGain) const
```

Comment:
This function is not used inside this class. So class user must call it before adding Exp. Is this proper?

Reply:
Just how the vanilla function is